src/sdc11073/wsdiscovery/wsdimpl.py

```diff
@@ -0,0 +1,620 @@
+        target_path_elements = [unquote(elem) for elem in target_path_elements]
+        if len(src_path_elements) > len(target_path_elements):
+            return False
+        return all(target_path_elements[i] == elem for i, elem in enumerate(src_path_elements))
+    if match_by == MATCH_BY_STRCMP:
+        return my_scope == other_scope
+    return False
+
+
+def match_type(type1: QName, type2: QName) -> bool:
+    """Check if namespace and localname are identical."""
+    return type1.namespace == type2.namespace and type1.localname == type2.localname
+
+
+def _is_type_in_list(ttype: QName, types: list[QName]) -> bool:
+    return any(match_type(ttype, entry) for entry in types)
+
+
+def _is_scope_in_list(uri: str, match_by: str, srv_sc: wsd_types.ScopesType) -> bool:
+    # returns True if every entry in scope.text is also found in srv_sc.text
+    # all entries are URIs
+    if srv_sc is None:
+        return False
+    return any(match_scope(uri, entry, match_by) for entry in srv_sc.text)
+
+
+def matches_filter(service: Service,
+                   types: list[QName] | None,
+                   scopes: wsd_types.ScopesType | None,
+                   logger: Logger | None = None) -> bool:
+    """Check if service matches the types and scopes."""
+    if types is not None and len(types) > 0:
+        srv_ty = service.types
+        for ttype in types:
+            if not _is_type_in_list(ttype, srv_ty):
+                if logger:
+                    logger.debug('types not matching: %r is not in types list %r', ttype, srv_ty)
+                return False
+        if logger:
+            logger.debug('matching types')
+    if scopes is not None:
+        srv_sc = service.scopes
+        for uri in scopes.text:
+            if not _is_scope_in_list(uri, scopes.MatchBy, srv_sc):
+                if logger:
+                    logger.debug('scope not matching: %s is not in scopes list %r', uri, srv_sc)
+                return False
+        if logger:
+            logger.debug('matching scopes')
+    return True
+
+
+def filter_services(services: Iterable[Service],
+                    types: list[QName] | None,
+                    scopes: wsd_types.ScopesType | None,
+                    logger: Logger | None = None) -> list[Service]:
+    """Filter services that match types and scopes."""
+    return [service for service in services if matches_filter(service, types, scopes, logger)]
+
+
+def generate_instance_id() -> str:
+    """Return a random number."""
+    return str(random.randint(1, 0xFFFFFFFF))
+
+
+def _mk_wsd_soap_message(header_info: HeaderInformationBlock,
+                         payload: MessageType) -> CreatedMessage:
+    # use discovery specific namespaces
+    return message_factory.mk_soap_message(header_info, payload,
+                                           ns_list=[nsh.S12, nsh.WSA, nsh.WSD], use_defaults=False)
+
+
+class WSDiscovery:

```

Comment:
make `WSDicovery` a context manager, such that it may be used like
```python
with WSDiscovery(...) as discovery:
    services = discovery.search_services(...)
```